src/vsr/replica.zig

```diff
@@ -5083,44 +5065,18 @@ pub fn ReplicaType(
-                    // This is the oldest op that is guaranteed to be in the WALs of any replica.
-                    // (Assuming that this primary has not been superseded.)
-                    const op_wal_oldest = @min(
-                        // Add the oldest pipeline_prepare_queue_max ops because they may have been
-                        // newer ops which were then truncated by a view-change, causing the head op
-                        // to backtrack.
-                        self.op + constants.pipeline_prepare_queue_max,
-                        // ...But the pipeline messages could not have moved past the next
-                        // checkpoint's op_prepare_max.
-                        self.op_prepare_max(),
-                    ) -| (constants.journal_slot_count - 1);
-
-                    // We know checkpoint ids for the previous checkpoint and the one before that.
-                    // Don't try repairing ops with older checkpoint_ids which are impossible to
-                    // verify.
-                    const op_with_checkpoint_id_oldest =
-                        (self.op_checkpoint() + 1) -| constants.vsr_checkpoint_interval;
-                    assert(self.checkpoint_id_for_op(op_with_checkpoint_id_oldest) != null);
-
-                    if (op_with_checkpoint_id_oldest > 0) {
-                        assert(self.checkpoint_id_for_op(op_with_checkpoint_id_oldest - 1) == null);
-                    }
+                if (self.op_checkpoint() == 0) {
+                    break :op 0;
+                }
 
-                    break :op @max(op_wal_oldest, op_with_checkpoint_id_oldest);
+                const op_checkpoint_trigger =
+                    vsr.Checkpoint.trigger_for_checkpoint(self.op_checkpoint()).?;
+                // After state sync, commit_max might lag behind checkpoint_op.
+                maybe(self.commit_max < op_checkpoint_trigger);
+                if (self.commit_max > op_checkpoint_trigger) {
+                    break :op self.op_checkpoint() + 1;
                 } else {
-                    // Strictly speaking a backup only needs to repair commit_min+1… to proceed.
-                    // However, if the backup crashes and recovers, it will need to replay ops
-                    // from the checkpoint, so we repair slightly more.
-                    if (self.op_checkpoint() == self.op) {
-                        // Don't allow "op_repair_min > op_head".
-                        break :op self.op_checkpoint();
-                    } else {
-                        // +1 because the primary (and other replicas) may overwrite op_checkpoint.
-                        // And we don't need op_checkpoint's prepare/header in our log if we crash.
-                        break :op self.op_checkpoint() + 1;
-                    }
+                    break :op (self.op_checkpoint() + 1) -| constants.vsr_checkpoint_interval;
                 }
             };
 
```

Comment:
Tried to add an extra assert here:

```
assert(self.prepare_max() - op < constants.journal_slot_count);
```

but that doesn't actually work for a funny reason:

* it can be violated if your `op_repair_min` still uses the previous checkpoint, and `prepare_max` already picks the next one (that is, commit_max is right on the trigger).
* but it actually isn't really violated --- when you receives that `op_prepare_max`, you advance your `commit_max`, which causes you to move op_repair_min by a checkpoint forward as well.